src/main/java/io/aiven/kafka/connect/opensearch/RecordConverter.java

```diff
@@ -130,7 +143,32 @@ private String getPayload(final SinkRecord record) {
                 : preProcessValue(record.value(), record.valueSchema(), schema);
 
         final byte[] rawJsonPayload = JSON_CONVERTER.fromConnectData(record.topic(), schema, value);
-        return new String(rawJsonPayload, StandardCharsets.UTF_8);
+        if (config.dataStreamEnabled()) {
+            return addTimestampToPayload(rawJsonPayload, record.timestamp());
+        } else {
+            return new String(rawJsonPayload, StandardCharsets.UTF_8);
+        }
+    }
+
+    private String addTimestampToPayload(final byte[] rawJsonPayload, final long timestamp) {
+        if (DATA_STREAM_TIMESTAMP_FIELD_DEFAULT.equals(config.dataStreamTimestampField())) {
+            try {
+                final var json = objectMapper.readTree(rawJsonPayload);
+                if (!json.isObject()) {
+                    throw new DataException(
+                            "JSON payload is a type of "
+                                    + json.getNodeType()
+                                    + ". Required is JSON Object.");
+                }
+                final var rootObject = (ObjectNode) json;
+                rootObject.put(DATA_STREAM_TIMESTAMP_FIELD_DEFAULT, timestamp);
+                return objectMapper.writeValueAsString(json);
```

Comment:
Once the https://github.com/aiven/opensearch-connector-for-apache-kafka/pull/158/files?diff=split&w=1#r1047568645 is in place, you could use overloaded version of IndexRequest::source without the need tp make the conversion back to string

Reply:
Same here lets address it in UPSERT PR?